File: apps/web/src/lib/integrations/race_match.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	haversineMetres,
	isRaceMatchCandidate,
	RACE_MATCH_RADIUS_M,
	RACE_MATCH_THRESHOLD,
	raceDistanceBand,
	raceMatchScore
} from './race_match';

test('raceDistanceBand buckets recognised distances', () => {
	assert.equal(raceDistanceBand(5000), '5k');
	assert.equal(raceDistanceBand(10000), '10k');
	assert.equal(raceDistanceBand(21097), 'half');
	assert.equal(raceDistanceBand(42195), 'marathon');
	assert.equal(raceDistanceBand(50000), 'ultra');
	assert.equal(raceDistanceBand(160934), 'ultra');
});

test('raceDistanceBand tolerance edges', () => {
	assert.equal(raceDistanceBand(4500), '5k');
	assert.equal(raceDistanceBand(5500), '5k');
	assert.equal(raceDistanceBand(9000), '10k');
	assert.equal(raceDistanceBand(11000), '10k');
});

test('raceDistanceBand returns null for unrecognised / tiny distances', () => {
	assert.equal(raceDistanceBand(4499), null);
	assert.equal(raceDistanceBand(7000), null); // between 5k and 10k bands
	assert.equal(raceDistanceBand(15000), null); // between 10k and half
	assert.equal(raceDistanceBand(null), null);
	assert.equal(raceDistanceBand(undefined), null);
	assert.equal(raceDistanceBand(NaN), null);
});

test('different calendar day scores 0', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-20T09:00:00Z', runStartLatLng: null, runDistanceM: 21097 },
		{ race_date: '2025-09-21', distance_m: 21097, distance_m_away: 100 }
	);
	assert.equal(score, 0);
});

test('same day with near start + matching band scores high', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21T09:00:00Z', runStartLatLng: { lat: 0, lng: 0 }, runDistanceM: 21097 },
		{ race_date: '2025-09-21', distance_m: 21097, distance_m_away: 200 }
	);
	assert.ok(score > 0.95, `expected near-1, got ${score}`);
});

test('same day only (no proximity, no band) normalises to 1', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21', runStartLatLng: null, runDistanceM: null },
		{ race_date: '2025-09-21', distance_m: null, distance_m_away: null }
	);
	// Only the day signal could be evaluated → 0.5 / 0.5 = 1.
	assert.equal(score, 1);
});

test('same day + matching band but no proximity still strong', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21', runStartLatLng: null, runDistanceM: 5000 },
		{ race_date: '2025-09-21', distance_m: 5000, distance_m_away: null }
	);
	// (0.5 + 0.2) / (0.5 + 0.2) = 1.
	assert.equal(score, 1);
});

test('same day + mismatched band lowers the score below threshold contribution', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21', runStartLatLng: null, runDistanceM: 5000 },
		{ race_date: '2025-09-21', distance_m: 42195, distance_m_away: null }
	);
	// (0.5) / (0.5 + 0.2) ≈ 0.714 — day matched, band did not.
	assert.ok(Math.abs(score - 0.5 / 0.7) < 1e-9, `got ${score}`);
});

test('proximity falloff is linear to the radius edge', () => {
	const half = raceMatchScore(
		{ runDate: '2025-09-21', runStartLatLng: { lat: 0, lng: 0 }, runDistanceM: null },
		{ race_date: '2025-09-21', distance_m: null, distance_m_away: RACE_MATCH_RADIUS_M / 2 }
	);
	// (0.5 + 0.3*0.5) / (0.5 + 0.3) = 0.65 / 0.8.
	assert.ok(Math.abs(half - 0.65 / 0.8) < 1e-9, `got ${half}`);
});

test('start beyond the radius contributes no proximity points', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21', runStartLatLng: { lat: 0, lng: 0 }, runDistanceM: null },
		{ race_date: '2025-09-21', distance_m: null, distance_m_away: RACE_MATCH_RADIUS_M * 3 }
	);
	// (0.5 + 0) / (0.5 + 0.3) = 0.625.
	assert.ok(Math.abs(score - 0.5 / 0.8) < 1e-9, `got ${score}`);
});

test('isRaceMatchCandidate gates on the threshold', () => {
	const near = {
		runDate: '2025-09-21',
		runStartLatLng: { lat: 0, lng: 0 },
		runDistanceM: 21097
	};
	assert.equal(
		isRaceMatchCandidate(near, { race_date: '2025-09-21', distance_m: 21097, distance_m_away: 100 }),
		true
	);
	assert.equal(
		isRaceMatchCandidate(near, { race_date: '2025-09-22', distance_m: 21097, distance_m_away: 100 }),
		false
	);
});

test('full ISO timestamp and date-only compare on the calendar day', () => {
	const score = raceMatchScore(
		{ runDate: '2025-09-21T23:59:00Z', runStartLatLng: null, runDistanceM: 10000 },
		{ race_date: '2025-09-21', distance_m: 10000, distance_m_away: null }
	);
	assert.equal(score, 1);
});

test('haversineMetres ~0 for identical points', () => {
	assert.ok(haversineMetres({ lat: 51.5, lng: -0.1 }, { lat: 51.5, lng: -0.1 }) < 1e-6);
});

test('haversineMetres ~111km per degree of latitude', () => {
	const d = haversineMetres({ lat: 0, lng: 0 }, { lat: 1, lng: 0 });
	assert.ok(Math.abs(d - 111195) < 500, `got ${d}`);
});

test('RACE_MATCH_THRESHOLD is the documented 0.5', () => {
	assert.equal(RACE_MATCH_THRESHOLD, 0.5);
});

File: apps/mobile_android/test/race_match_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/race_match.dart';

void main() {
  group('raceDistanceBand', () {
    test('buckets recognised distances', () {
      expect(raceDistanceBand(5000), RaceDistanceBand.fiveK);
      expect(raceDistanceBand(10000), RaceDistanceBand.tenK);
      expect(raceDistanceBand(21097), RaceDistanceBand.half);
      expect(raceDistanceBand(42195), RaceDistanceBand.marathon);
      expect(raceDistanceBand(50000), RaceDistanceBand.ultra);
      expect(raceDistanceBand(160934), RaceDistanceBand.ultra);
    });

    test('tolerance edges', () {
      expect(raceDistanceBand(4500), RaceDistanceBand.fiveK);
      expect(raceDistanceBand(5500), RaceDistanceBand.fiveK);
      expect(raceDistanceBand(9000), RaceDistanceBand.tenK);
      expect(raceDistanceBand(11000), RaceDistanceBand.tenK);
    });

    test('returns null for unrecognised / tiny distances', () {
      expect(raceDistanceBand(4499), isNull);
      expect(raceDistanceBand(7000), isNull);
      expect(raceDistanceBand(15000), isNull);
      expect(raceDistanceBand(null), isNull);
      expect(raceDistanceBand(double.nan), isNull);
    });
  });

  group('raceMatchScore', () {
    test('different calendar day scores 0', () {
      final s = raceMatchScore(
        const RunMatchInput(
            runDate: '2025-09-20T09:00:00Z', runStartLatLng: null, runDistanceM: 21097),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: 21097, distanceMAway: 100),
      );
      expect(s, 0);
    });

    test('same day with near start + matching band scores high', () {
      final s = raceMatchScore(
        const RunMatchInput(
            runDate: '2025-09-21T09:00:00Z',
            runStartLatLng: LatLng(0, 0),
            runDistanceM: 21097),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: 21097, distanceMAway: 200),
      );
      expect(s, greaterThan(0.95));
    });

    test('same day only (no proximity, no band) normalises to 1', () {
      final s = raceMatchScore(
        const RunMatchInput(runDate: '2025-09-21', runStartLatLng: null, runDistanceM: null),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: null, distanceMAway: null),
      );
      expect(s, 1);
    });

    test('same day + matching band but no proximity still strong', () {
      final s = raceMatchScore(
        const RunMatchInput(runDate: '2025-09-21', runStartLatLng: null, runDistanceM: 5000),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: 5000, distanceMAway: null),
      );
      expect(s, 1);
    });

    test('same day + mismatched band lowers the score', () {
      final s = raceMatchScore(
        const RunMatchInput(runDate: '2025-09-21', runStartLatLng: null, runDistanceM: 5000),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: 42195, distanceMAway: null),
      );
      expect((s - 0.5 / 0.7).abs() < 1e-9, isTrue);
    });

    test('proximity falloff is linear to the radius edge', () {
      final s = raceMatchScore(
        const RunMatchInput(
            runDate: '2025-09-21', runStartLatLng: LatLng(0, 0), runDistanceM: null),
        const ListingMatchInput(
            raceDate: '2025-09-21', distanceM: null, distanceMAway: raceMatchRadiusM / 2),
      );
      expect((s - 0.65 / 0.8).abs() < 1e-9, isTrue);
    });

    test('start beyond the radius contributes no proximity points', () {
      final s = raceMatchScore(
        const RunMatchInput(
            runDate: '2025-09-21', runStartLatLng: LatLng(0, 0), runDistanceM: null),
        const ListingMatchInput(
            raceDate: '2025-09-21', distanceM: null, distanceMAway: raceMatchRadiusM * 3),
      );
      expect((s - 0.5 / 0.8).abs() < 1e-9, isTrue);
    });

    test('full ISO timestamp compares on the calendar day', () {
      final s = raceMatchScore(
        const RunMatchInput(
            runDate: '2025-09-21T23:59:00Z', runStartLatLng: null, runDistanceM: 10000),
        const ListingMatchInput(raceDate: '2025-09-21', distanceM: 10000, distanceMAway: null),
      );
      expect(s, 1);
    });
  });

  group('isRaceMatchCandidate', () {
    test('gates on the threshold', () {
      const near = RunMatchInput(
          runDate: '2025-09-21', runStartLatLng: LatLng(0, 0), runDistanceM: 21097);
      expect(
        isRaceMatchCandidate(near,
            const ListingMatchInput(raceDate: '2025-09-21', distanceM: 21097, distanceMAway: 100)),
        isTrue,
      );
      expect(
        isRaceMatchCandidate(near,
            const ListingMatchInput(raceDate: '2025-09-22', distanceM: 21097, distanceMAway: 100)),
        isFalse,
      );
    });
  });

  group('haversineMetres', () {
    test('~0 for identical points', () {
      expect(haversineMetres(const LatLng(51.5, -0.1), const LatLng(51.5, -0.1)) < 1e-6, isTrue);
    });

    test('~111km per degree of latitude', () {
      final d = haversineMetres(const LatLng(0, 0), const LatLng(1, 0));
      expect((d - 111195).abs() < 500, isTrue);
    });
  });

  test('raceMatchThreshold is the documented 0.5', () {
    expect(raceMatchThreshold, 0.5);
  });
}
